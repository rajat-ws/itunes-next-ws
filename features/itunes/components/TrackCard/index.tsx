import { useRef, useState } from "react";
import { useRouter } from "next/router";
import styled from "styled-components";
import isEmpty from "lodash/isEmpty";
import { Card, Image, Typography } from "antd";
import { colors, fonts } from "@app/themes";
import { If } from "@app/common";
import { TrackItem } from "../../api/getTracks";

const { Paragraph } = Typography;

export type TrackCardProps = {
  data: TrackItem;
  isShowDetailsButton: boolean;
  isShowDetails: boolean;
  handlePlayPauseWrapper;
};

const TrackCardWrapper = styled(Card)`
  && {
    border-radius: 0.5rem;
    width: 25rem;
    border: 1px solid ${colors.secondary};
    text-align: center;
    border: 1px solid ${colors.primary}
`;

const StyledDescription = styled.div`
  display: flex;
  flex-direction: column;
  ${fonts.size.big}
  padding-top: 1rem;
`;

const StyledSpan = styled.span`
  && {
    color: ${colors.text};
    ${fonts.weights.bold}
    ${fonts.size.big}
  }
`;

const ButtonWrapper = styled.div`
  && {
    display: flex;
    flex-direction: column;
    gap: 2rem;
  }
`;

const ShowDetailsButton = styled.button`
  && {
    background-color: ${colors.primary};
    color: ${colors.textSecondary};
    padding: 5px;
    cursor: pointer;
    border: none;
    border-radius: 0.5rem;
    width: 100%;
    ${fonts.size.xRegular}
    ${fonts.weights.bold}

    &:hover {
      background-color: ${colors.primaryDark};
    }
  }
`;

const PlayButton = styled.button`
  && {
    background-color: ${colors.success};
    color: ${colors.textSecondary};
    border: none;
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: 0.5rem;
    width: 100%;
    cursor: pointer;
    &:hover {
      background-color: ${colors.successLight};
    }
  }
`;

const ButtonLabel = styled.span`
  && {
    ${fonts.weights.bold}
    ${fonts.size.big}
  }
`;

const TrackCard: React.FC<TrackCardProps> = ({
  data,
  isShowDetailsButton,
  isShowDetails,
  handlePlayPauseWrapper,
}) => {
  const {
    artistName,
    trackName,
    trackId,
    trackTimeMillis: trackDuration,
    artworkUrl100: imageUrl,
    previewUrl,
    collectionName,
    country,
    primaryGenreName,
    kind,
    wrapperType,
  } = data;
  const router = useRouter();
  const [isTrackPlaying, setIsTrackPlaying] = useState<boolean>(false);

  const audioRef = useRef<HTMLAudioElement>(null);

  const handlePlayPause = () => {
    const isTrackPaused = audioRef.current?.paused;

    if (isTrackPaused) {
      audioRef.current.play();
    } else {
      audioRef.current?.pause();
    }
    setIsTrackPlaying(!isTrackPlaying);

    if (!isShowDetails) {
      handlePlayPauseWrapper(audioRef);
    }
  };

  return (
    <TrackCardWrapper data-testid="track-card">
      <If condition={!isEmpty(imageUrl)} otherwise={"No image available"}>
        <Image src={imageUrl} width="80%" alt={collectionName} />
      </If>
      <StyledDescription>
        <If
          condition={!isEmpty(artistName)}
          otherwise={<Paragraph>No artist name available</Paragraph>}
        >
          <Paragraph>
            <StyledSpan> Artist name: </StyledSpan> {artistName}
          </Paragraph>
        </If>
        <If
          condition={!isEmpty(primaryGenreName)}
          otherwise={<Paragraph>No primaryGenre name available</Paragraph>}
        >
          <Paragraph>
            <StyledSpan> Genre: </StyledSpan> {primaryGenreName}
          </Paragraph>
        </If>

        <If
          condition={!isEmpty(trackName)}
          otherwise={
            <Paragraph>
              <StyledSpan> Track name: </StyledSpan> No track name available
            </Paragraph>
          }
        >
          <Paragraph>
            <StyledSpan> Track name: </StyledSpan> {trackName}
          </Paragraph>
        </If>

        <If condition={isShowDetails}>
          <Paragraph>
            <StyledSpan> Country: </StyledSpan> {country}
          </Paragraph>
          <Paragraph>
            <StyledSpan> Kind: </StyledSpan> {kind}
          </Paragraph>

          <Paragraph>
            <StyledSpan> Collection name: </StyledSpan> {collectionName}
          </Paragraph>

          <Paragraph>
            <StyledSpan> Wrapper Type: </StyledSpan> {wrapperType}
          </Paragraph>

          {trackDuration && (
            <Paragraph>
              <StyledSpan> Duration: </StyledSpan> {Math.floor(trackDuration / 60000)}:
              {Math.floor(trackDuration / 1000) % 60}s
            </Paragraph>
          )}
        </If>
      </StyledDescription>

      <ButtonWrapper>
        <If condition={isShowDetailsButton}>
          <ShowDetailsButton
            data-testid="showDetails"
            onClick={() => router.push(`/trackDetails/${trackId}`)}
          >
            Show Details
          </ShowDetailsButton>
        </If>
        <PlayButton onClick={handlePlayPause} data-testid="playPauseBtn">
          <If
            condition={audioRef.current?.src && !audioRef.current?.paused}
            otherwise={<ButtonLabel> Play </ButtonLabel>}
          >
            <ButtonLabel> Pause </ButtonLabel>
          </If>
        </PlayButton>
      </ButtonWrapper>
      <audio src={previewUrl} data-testid="trackAudio" ref={audioRef} />
    </TrackCardWrapper>
  );
};

export default TrackCard;
