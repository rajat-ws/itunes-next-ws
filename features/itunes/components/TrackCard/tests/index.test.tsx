import { render } from "@app/utils/testUtils";
import TrackCard from "../index";

describe("<TrackCard />", () => {
  const trackCardProps = {
    data: {
      trackId: 12345,
      artistName: "Anuv",
      collectionName: "Collection",
      trackName: "Baarishein",
      artworkUrl100:
        "https://is5-ssl.mzstatic.com/image/thumb/Music116/v4/5a/47/6d/5a476ddd-4690-1297-1896-6a286a497a21/191404113974.png/100x100bb.jpg",
    },
  };

  it("should render and match the snapsho t", () => {
    const { baseElement } = render(<TrackCard {...trackCardProps} />);
    expect(baseElement).toMatchSnapshot();
  });
});
