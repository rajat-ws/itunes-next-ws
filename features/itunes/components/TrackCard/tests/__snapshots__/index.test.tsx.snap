// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered TrackCard__TrackCardWrapper-sc-k4gp1w-0 fuDzyM"
      data-testid="track-card"
    >
      <div
        class="ant-card-body"
      >
        <div
          class="ant-image"
          style="width: 80%;"
        >
          <img
            alt="collecition"
            class="ant-image-img"
            src="https://is5-ssl.mzstatic.com/image/thumb/Music116/v4/5a/47/6d/5a476ddd-4690-1297-1896-6a286a497a21/191404113974.png/100x100bb.jpg"
          />
          <div
            class="ant-image-mask"
          >
            <div
              class="ant-image-mask-info"
            >
              <span
                aria-label="eye"
                class="anticon anticon-eye"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="eye"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
                  />
                </svg>
              </span>
              Preview
            </div>
          </div>
        </div>
        <div
          class="TrackCard__StyledDescription-sc-k4gp1w-1 yMRaE"
        >
          <div
            class="ant-typography"
          >
            <span
              class="TrackCard__StyledSpan-sc-k4gp1w-2 kMUszr"
            >
               Artist name: 
            </span>
             
            Adele
          </div>
          <div
            class="ant-typography"
          >
            <span
              class="TrackCard__StyledSpan-sc-k4gp1w-2 kMUszr"
            >
               Collection name: 
            </span>
             
            collecition
          </div>
          <div
            class="ant-typography"
          >
            <span
              class="TrackCard__StyledSpan-sc-k4gp1w-2 kMUszr"
            >
               Track name: 
            </span>
             
            Hello
          </div>
          <div
            class="ant-typography"
          >
            <span
              class="TrackCard__StyledSpan-sc-k4gp1w-2 kMUszr"
            >
               Country: 
            </span>
             
            India
          </div>
          <div
            class="ant-typography"
          >
            <span
              class="TrackCard__StyledSpan-sc-k4gp1w-2 kMUszr"
            >
               Kind: 
            </span>
             
            Kind
          </div>
          <div
            class="ant-typography"
          >
            <span
              class="TrackCard__StyledSpan-sc-k4gp1w-2 kMUszr"
            >
               Genre: 
            </span>
             
            abc
          </div>
          <div
            class="ant-typography"
          >
            <span
              class="TrackCard__StyledSpan-sc-k4gp1w-2 kMUszr"
            >
               Wrapper Type: 
            </span>
             
            Rap
          </div>
          <div
            class="ant-typography"
          >
            <span
              class="TrackCard__StyledSpan-sc-k4gp1w-2 kMUszr"
            >
               Duration: 
            </span>
             
            4
            :
            55
            s
          </div>
        </div>
        <div
          class="TrackCard__ButtonWrapper-sc-k4gp1w-3 ilHQNx"
        >
          <button
            class="TrackCard__PlayButton-sc-k4gp1w-5 jELdXO"
            data-testid="playPauseBtn"
          >
            <span
              class="TrackCard__ButtonLabel-sc-k4gp1w-6 shsTk"
            >
               Play 
            </span>
          </button>
        </div>
        <audio
          data-testid="trackAudio"
          src="22233223"
        />
      </div>
    </div>
  </div>
</body>
`;
